plugins/filter_aws/aws.c

```diff
@@ -370,6 +400,208 @@ static int get_vpc_metadata(struct flb_filter_aws *ctx)
     return ret;
 }
 
+void flb_filter_aws_tags_destroy(struct flb_filter_aws *ctx)
+{
+    size_t i;
+    if (!ctx) {
+        return;
+    }
+    if (ctx->tag_keys) {
+        for (i = 0; i < ctx->tags_count; i++) {
+            if (ctx->tag_keys[i]) {
+                flb_sds_destroy(ctx->tag_keys[i]);
+            }
+        }
+        flb_free(ctx->tag_keys);
+        ctx->tag_keys = NULL;
+    }
+    if (ctx->tag_values) {
+        for (i = 0; i < ctx->tags_count; i++) {
+            if (ctx->tag_values[i]) {
+                flb_sds_destroy(ctx->tag_values[i]);
+            }
+        }
+        flb_free(ctx->tag_values);
+        ctx->tag_values = NULL;
+    }
+    if (ctx->tag_keys_len) {
+        flb_free(ctx->tag_keys_len);
+    }
+    ctx->tag_keys_len = NULL;
+    if (ctx->tag_values_len) {
+        flb_free(ctx->tag_values_len);
+    }
+    ctx->tag_values_len = NULL;
+    ctx->tags_count = 0;
+}
+
+/* Get EC2 instance tag keys from /latest/meta-data/tags/instance.
+ * Initializes ctx->tags_count, ctx->tag_keys and ctx->tag_keys_len.
+ */
+static int get_ec2_tag_keys(struct flb_filter_aws *ctx)
+{
+    int ret;
+    flb_sds_t tags_list = NULL;
+    size_t len = 0;
+    size_t tag_index = 0;
+    size_t tag_start = 0;
+    size_t tag_end = 0;
+    flb_sds_t tag_key;
+    size_t tag_key_len;
+    size_t i;
+
+    /* get a list of tag keys from the meta data server */
+    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_INSTANCE_TAG, &tags_list, &len);
+    if (ret < 0) {
+        if (ret == -2) { /* if there are no tags, response status code is 404 */
+            flb_plg_warn(ctx->ins, "tags not available in the EC2 instance metadata");
+            ctx->tags_count = 0;
+            return 0;
+        }
+        flb_sds_destroy(tags_list);
+        return -1;
+    }
+
+    /* count number of tag keys and allocate memory for pointers and lengths */
+    /* since get_metadata returned 0, we assume there is at least 1 tag */
+    /* \n is separator, therefore number of items = number of \n + 1 */
```

Comment:
How do you handle the case where the string is empty and has no tags? Or is that impossible?

I'd feel safer if that case was explicitly handled or noted in a comment if its impossible.

Reply:
> How do you handle the case where the string is empty and has no tags? Or is that impossible?

I don't think it's possible to receive an empty response. At least I wasn't able to get such response during my testing.

> I'd feel safer if that case was explicitly handled or noted in a comment if its impossible.

Sure, I agree.
Done, https://github.com/fluent/fluent-bit/pull/6065/commits/d6d20e76a9787c7df041a922820c012f53de125e